File: rxhttp/src/main/java/com/senierr/http/builder/HeaderBuilder.java
package com.senierr.http.builder;

import java.util.LinkedHashMap;

import io.reactivex.annotations.NonNull;
import okhttp3.Headers;

/**
 * Http请求头构建器
 *
 * @author zhouchunjie
 * @date 2018/8/29
 */
public final class HeaderBuilder {

    private LinkedHashMap<String, String> headers = new LinkedHashMap<>();

    public void addHeader(@NonNull String key, @NonNull String value) {
        headers.put(key, value);
    }

    public void addHeaders(@NonNull LinkedHashMap<String, String> headers) {
        this.headers.putAll(headers);
    }

    public @NonNull LinkedHashMap<String, String> getHeaders() {
        return headers;
    }

    public @NonNull Headers build(){
        Headers.Builder builder = new Headers.Builder();
        for (String key: headers.keySet()) {
            String value = headers.get(key);
            if (value != null && value.length() > 0) {
                builder.add(key, value);
            }
        }
        return builder.build();
    }
}

File: rxhttp/src/main/java/com/senierr/http/builder/HeaderBuilder.kt
package com.senierr.http.builder

import okhttp3.Headers

/**
 * Http请求头构建器
 *
 * @author zhouchunjie
 * @date 2018/8/29
 */
class HeaderBuilder : Builder<Headers> {

    private val baseHeaders = LinkedHashMap<String, String>()
    private var ignoreBaseHeaders = false

    private val headers = LinkedHashMap<String, String>()

    fun addBaseHeader(key: String, value: String) {
        baseHeaders[key] = value
    }

    fun addBaseHeaders(headers: LinkedHashMap<String, String>) {
        baseHeaders.putAll(headers)
    }

    fun ignoreBaseHeaders() {
        ignoreBaseHeaders = true
    }

    fun addHeader(key: String, value: String) {
        headers[key] = value
    }

    fun addHeaders(headers: LinkedHashMap<String, String>) {
        this.headers.putAll(headers)
    }

    override fun build(): Headers {
        // 若设置了基础头，且未忽略，添加基础头（注意添加顺序）
        val actualHeaders = LinkedHashMap<String, String>()
        if (baseHeaders.isNotEmpty() && !ignoreBaseHeaders) {
            actualHeaders.putAll(baseHeaders)
        }
        actualHeaders.putAll(headers)

        val builder = Headers.Builder()
        for (key in actualHeaders.keys) {
            val value = actualHeaders[key]
            value?.let { builder.add(key, it) }
        }
        return builder.build()
    }
}